File: react/src/main/java/com/react/auth/controller/AuthenticationController.java
package com.react.auth.controller;

import com.react.auth.dto.*;
import com.react.auth.service.AuthenticationService;
import com.react.auth.service.JwtService;
import com.react.auth.service.RefreshTokenService;
import com.react.model.user.UserRole;
import jakarta.validation.Valid;
import lombok.RequiredArgsConstructor;
import org.springframework.http.ResponseEntity;
import org.springframework.security.core.userdetails.UserDetails;
import org.springframework.security.core.userdetails.UserDetailsService;
import org.springframework.web.bind.annotation.*;

@RestController
@RequestMapping("/api/auth")
@RequiredArgsConstructor
public class AuthenticationController {

    private final AuthenticationService authenticationService;
    private final RefreshTokenService refreshTokenService;
    private final JwtService jwtService;
    private final UserDetailsService userDetailsService;

    @PostMapping("/register")
    public ResponseEntity<AuthenticationResponse> register(
            @RequestBody RegisterRequest request
    ) {
        return ResponseEntity.ok(authenticationService.register(request));
    }

    @PostMapping("/authenticate")
    public ResponseEntity<AuthenticationResponse> authenticate(
            @Valid @RequestBody AuthenticationRequest request
    ) {
        return ResponseEntity.ok(authenticationService.authenticate(request));
    }

    @PostMapping("/refresh")
    public ResponseEntity<AuthenticationResponse> refresh(
            @RequestBody RefreshRequest req
    ) {

        var rotated = refreshTokenService.rotate(req.getRefreshToken());

        UserDetails userDetails =
                userDetailsService.loadUserByUsername(
                        rotated.getUserId().toString()
                );

        String access = jwtService.generateToken(userDetails);

        return ResponseEntity.ok(
                new AuthenticationResponse(
                        access,
                        UserRole.STUDENT,
                        rotated.getToken()
                )
        );
    }

    @GetMapping("/logout")
    public ResponseEntity<?> logout(
            @RequestBody LogoutRequest request
    ) {
        refreshTokenService.revokeByToken(request.getRefreshToken());
        return ResponseEntity.ok().build();
    }
}

File: BaseSpringSecurity/src/main/kotlin/auth/controller/AuthenticationController.kt
package com.auth.controller

import com.auth.payload.AuthenticationRequest
import com.auth.payload.AuthenticationResponse
import com.auth.service.AuthenticationService
import com.auth.payload.RegisterRequest
import com.auth.service.RefreshTokenService
import jakarta.validation.Valid
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RestController
import com.auth.dto.RefreshRequest
import com.auth.service.JwtService
import org.springframework.security.core.userdetails.UserDetailsService


@RestController
@RequestMapping("/api/auth")
open class AuthenticationController(
    private val authenticationService: AuthenticationService,
    private val refreshTokenService: RefreshTokenService,
    private val jwtService: JwtService,
    private val userDetailService: UserDetailsService
) {

    @PostMapping("/register")
    fun register(
        @RequestBody request: RegisterRequest
    ): ResponseEntity<AuthenticationResponse> {
        return ResponseEntity.ok(authenticationService.register(request))
    }

    @PostMapping("/authenticate")
    open fun authenticate(
        @Valid
        @RequestBody request: AuthenticationRequest,
    ): ResponseEntity<AuthenticationResponse> {
        println("vào được controller")
        return ResponseEntity.ok(authenticationService.authenticate(request))
    }

    @PostMapping("/refresh")
    open fun refresh(@RequestBody req: RefreshRequest): ResponseEntity<AuthenticationResponse> {
        println("refresh")
        val rotated = refreshTokenService.rotate(req.refreshToken)
        val userDetails = userDetailService.loadUserByUsername(rotated.userId.toString())
        val access = jwtService.generateToken(userDetails)
        return ResponseEntity.ok(AuthenticationResponse(access,
            "STUDENT",
            rotated.token))
    }

}